### PR TITLE
rule manager: Moved metric registration to custom registerer which is already available.

### PR DESCRIPTION
**Pull request by bwplotka**
Hi :wave:  

This is to start discussion. We started notion of custom registerer in this package for some metrics calculated by manager `Collect` method itself, but we did not go full way with out other metrics.

Since in Thanos we don't use global registry, can we use Manager registry to register those as well?

TODO in this PR, move metrics itself to some metric struct owned by Manager itself.
  
Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

### CODE REVIEW

**Comment by brancz** [comment]
> TODO in this PR, move metrics itself to some metric struct owned by Manager itself.

What benefit does that have?

Otherwise this looks good to me.

**Comment by brian-brazil** [comment]
> What benefit does that have?

The PR would not be correct otherwise, as you'd be sharing metrics across managers.

**Comment by bwplotka** [comment]
> The PR would not be correct otherwise, as you'd be sharing metrics across managers.

exactly.

However creating multiple managers with same register will still return error (panic) from register, but I think that's expected, right?

**Comment by brian-brazil** [comment]
Yes, that's expected.

**Comment by brancz** [comment]
Whops. I'm so used to only having the descriptions as vars these days, that I read this code the way you meant it to be. In that case your suggestion makes sense, let's change it to do that.

**Comment by brian-brazil** [comment]
You still need to make that change.

**Comment by bwplotka** [comment]
I am aware, thanks for ping - just moving all of those to `prometheus.NewDesc(` is quite tedious (: lot's of setters and getters needed for `RuleGroup` wonder if we can avoid this. Probably not.

File: rules/manager.go [hunk 1]
```diff
@@ -110,12 +61,85 @@ var (
...
-	prometheus.MustRegister(evalFailures)
-	prometheus.MustRegister(evalDuration)
+type groupMetrics struct {
+	evalDuration        prometheus.Summary
+	evalFailures        prometheus.Counter
+	evalTotal           prometheus.Counter
+	iterationDuration   prometheus.Summary
+	iterationsMissed    prometheus.Counter
+	iterationsScheduled prometheus.Counter
+	lastEvaluation      *prometheus.GaugeVec
```
Comment (brian-brazil) [inline rules/manager.go]: Why are these pointers and the others not?

File: rules/manager.go [hunk 2]
```diff
@@ -307,34 +336,6 @@ func (g *Group) hash() uint64 {
 	return l.Hash()
 }
 
-// GetEvaluationDuration returns the time in seconds it took to evaluate the rule group.
-func (g *Group) GetEvaluationDuration() time.Duration {
```
Comment (brian-brazil) [inline rules/manager.go]: These are used in the UI

File: rules/manager.go (hunk 1 above)
Comment (bwplotka) [reply rules/manager.go]: This is struct, rest are interfaces.

File: rules/manager.go (hunk 2 above)
Comment (bwplotka) [reply rules/manager.go]: Hm... why CI is green then?

Comment (brian-brazil) [reply rules/manager.go]: CI doesn't test those.

**Comment by bwplotka** [comment]
Ok, reverted interface change, but things I don't really like and most likely needs to be fixed in this PR:
* Group is used like rule but not. No one ever use it as `Rule` interface, however it tries to implement all boilerplate methods. It is super misleading.
* No one knows what methods UI uses, so you never know until you manually run UI and check - I think some unit test for compatibility against `rules.html` template is must have.

**Comment by brian-brazil** [comment]
Thanks!

> I think some unit test for compatibility against rules.html template is must have.

That's been the case for a while.

**Pull request merged**